File: Greedy Algorithms and Their Applications/Fractional Knapsack Problem (Greedy Algorithm)/Fractional-Knapsack-Problem.java
import java.util.Arrays;
import java.util.Scanner;

class Item {
    int value, weight;

    Item(int value, int weight) {
        this.value = value;
        this.weight = weight;
    }
}

public class FractionalKnapsack {
    // Comparator for sorting items by value-to-weight ratio
    static void sortItemsByRatio(Item[] items) {
        Arrays.sort(items, (a, b) -> Double.compare((double) b.value / b.weight, (double) a.value / a.weight));
    }

    static double fractionalKnapsack(int capacity, Item[] items) {
        sortItemsByRatio(items); // Sort in descending order
        double totalValue = 0.0;

        for (Item item : items) {
            if (capacity >= item.weight) {
                capacity -= item.weight;
                totalValue += item.value;
            } else {
                totalValue += item.value * ((double) capacity / item.weight);
                break;
            }
        }

        return totalValue;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter number of items: ");
        int n = scanner.nextInt();

        Item[] items = new Item[n];
        System.out.println("Enter value and weight for each item:");
        for (int i = 0; i < n; i++) {
            int value = scanner.nextInt();
            int weight = scanner.nextInt();
            items[i] = new Item(value, weight);
        }

        System.out.print("Enter knapsack capacity: ");
        int capacity = scanner.nextInt();

        System.out.printf("Maximum value in Knapsack: %.2f%n", fractionalKnapsack(capacity, items));

        scanner.close();
    }
}

File: Greedy Algorithms and Their Applications/Fractional Knapsack Problem (Greedy Algorithm)/Fractional-Knapsack-Problem.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int value, weight;
} Item;

// Comparison function for sorting by value/weight ratio
int compare(const void *a, const void *b) {
    double ratio1 = ((Item *)a)->value / (double)((Item *)a)->weight;
    double ratio2 = ((Item *)b)->value / (double)((Item *)b)->weight;
    return (ratio2 > ratio1) - (ratio1 > ratio2); // Descending order
}

double fractional_knapsack(int capacity, Item items[], int n) {
    qsort(items, n, sizeof(Item), compare); // Sort items by value/weight ratio
    double total_value = 0.0;

    for (int i = 0; i < n; i++) {
        if (capacity >= items[i].weight) {
            capacity -= items[i].weight;
            total_value += items[i].value;
        } else {
            total_value += items[i].value * ((double)capacity / items[i].weight);
            break;
        }
    }

    return total_value;
}

int main() {
    int n, capacity;
    
    printf("Enter number of items: ");
    scanf("%d", &n);

    Item items[n];
    printf("Enter value and weight for each item:\n");
    for (int i = 0; i < n; i++) {
        scanf("%d %d", &items[i].value, &items[i].weight);
    }

    printf("Enter knapsack capacity: ");
    scanf("%d", &capacity);

    printf("Maximum value in Knapsack: %.2f\n", fractional_knapsack(capacity, items, n));
    return 0;
}
